core/src/modules/DynamicsModule/BBMDynamics.cpp

```diff
@@ -0,0 +1,100 @@
+ */
+
+#include "include/BBMDynamics.hpp"
+
+#include "include/gridNames.hpp"
+
+namespace Nextsim {
+
+static const std::vector<std::string> namedFields = { hiceName, ciceName, uName, vName };
+BBMDynamics::BBMDynamics()
+    : IDynamics(true)
+    , kernel(params)
+{
+    getStore().registerArray(Protected::ICE_U, &uice, RO);
+    getStore().registerArray(Protected::ICE_V, &vice, RO);
+}
+
+void BBMDynamics::setData(const ModelState::DataMap& ms)
+{
+    IDynamics::setData(ms);
+
+    bool isSpherical;
+    // Decide between Cartesian (x & y) and spherical (longitude & latitude)
+    if (ms.count(longitudeName) > 0 && ms.count(latitudeName) > 0) {
+        isSpherical = true;
+    } else if (ms.count(xName) > 0 && ms.count(yName) > 0) {
+        isSpherical = false;
+    } else {
+        // Throw a runtime_error exception which can either be handled or not
+        throw std::runtime_error(
+                "Input data must contain either Cartesian (" + xName + ", " + yName
+                        + ") or spherical (" + longitudeName + ", " + latitudeName
+                        + ") coordinates.");
+    }
+
+    // TODO: Remove this when spherical coordinates are fully implemented
+    if (isSpherical) std::cout << "Spherical coordinates are not yet implemented. Reverting to Cartesian." << std::endl;
+    isSpherical = false;
+    ModelArray fake25kmCoords(ModelArray::Type::VERTEX);
+    double d = 25000; // 25 km in metres
+    // Fill the fake coordinate array
+    for (size_t j = 0; j < ModelArray::size(ModelArray::Dimension::YVERTEX); ++j) {
+        for (size_t i = 0; i < ModelArray::size(ModelArray::Dimension::XVERTEX); ++i) {
+            fake25kmCoords.components({i, j})[0] = d * i;
+            fake25kmCoords.components({i, j})[1] = d * j;
+        }
+    }
+    ModelArray& coords = fake25kmCoords;
+    // End of code to be removed
+
+    // ModelArray& coords = ms.at(coordsName);
+    // TODO: Some encoding of the periodic edge boundary conditions
+    kernel.initialise(coords, isSpherical, ms.at(maskName));
+
+    uice = ms.at(uName);
+    vice = ms.at(vName);
+
+    // Set the data in the kernel arrays.
+    for (const auto& fieldName : namedFields) {
+        kernel.setData(fieldName, ms.at(fieldName));
+    }
+}
+
+void BBMDynamics::update(const TimestepTime& tst)
+{
+    std::cout << tst.start << std::endl;
+
+    // Fill the updated damage array with the initial value
+    damage = damage0.data();
+
+    // set the updated ice thickness, concentration and damage
+    kernel.setData(hiceName, hice.data());
+    kernel.setData(ciceName, cice.data());
+    kernel.setData(damageName, damage);
+
+    // set the forcing velocities
+    kernel.setData(uWindName, uwind.data());
+    kernel.setData(vWindName, vwind.data());
+    kernel.setData(uOceanName, uocean.data());
+    kernel.setData(vOceanName, vocean.data());
+
+    //kernel.setData(uName, uice);
+    //kernel.setData(vName, vice);
```

Comment:
The definitive arrays that hold the ice velocity components are the arrays in the dynamics, currently specifically the `CGVector` arrays in `CGDynamicsKernel`. Since there is (currently) nothing that changes the velocity outside the dynamics, the ice velocities are not set here, as the definitive values already reside in the dynamics kernel. Note that after the update the DG0/finite volume term of the ice velocity is copied back out to be accessible by the model outside the ice dynamics.

Reply:
I have added a comment to this effect and deleted the commented out code.